### PR TITLE
feat(platform/gitea): autodiscover repos by namespace

### PR DESCRIPTION
**Pull request by rvanbutselaar**
## Changes

Adds Gitea support for the configuration option autodiscoverNamespaces to run Renovate against all repositories in a namespace (Gitea organization).

## Context

Closes https://github.com/renovatebot/renovate/discussions/26792

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/platform/gitea/index.ts [hunk 1]
```diff
@@ -175,6 +175,13 @@ async function fetchRepositories(topic?: string): Promise<string[]> {
   return repos.filter((r) => !r.mirror).map((r) => r.full_name);
 }
 
+async function fetchOrgRepositories(organization: string): Promise<string[]> {
+  const repos = await helper.orgListRepos(organization);
+
+  // filter archived repositories because Gitea orgListRepos does not support this.
+  return repos.filter((r) => !r.mirror && !r.archived).map((r) => r.full_name);
+}
+
```
[the remark this answers is not in the document]
Comment (rvanbutselaar) [reply lib/modules/platform/gitea/index.ts]: The https://try.gitea.io/api/swagger#/organization/orgListRepos endpoint does not support filtering using topcis.

**Review by viceice — COMMENTED** [review]
please add a note for Gitea/ Forgejo that topics can't be used with namespace filter and that topics are preferred and namespace filter ignored.

**Comment by rvanbutselaar** [comment]
Added a note to: docs/usage/self-hosted-configuration.md

File: lib/modules/platform/gitea/index.ts [hunk 2]
```diff
@@ -319,12 +319,25 @@ const platform: Platform = {
   async getRepos(config?: AutodiscoverConfig): Promise<string[]> {
     logger.debug('Auto-discovering Gitea repositories');
     try {
-      if (!config?.topics) {
+      if (config?.topics) {
+        logger.debug('Auto-discovering Gitea repositories by topics');
```
Comment (viceice) [inline lib/modules/platform/gitea/index.ts:323]: @rarkins should we log topics here? just in case force easier trouble shooting?

Comment (rvanbutselaar) [reply lib/modules/platform/gitea/index.ts:323]: @viceice let me know if we need to create additional logging

Comment (viceice) [reply lib/modules/platform/gitea/index.ts:323]:
```suggestion
        logger.debug({ topics: config.topics }, 'Auto-discovering by topics');
```

File: lib/modules/platform/gitea/index.ts [hunk 3]
```diff
@@ -319,12 +319,25 @@ const platform: Platform = {
...
+        const repos = await map(config.topics, fetchRepositories);
+        return deduplicateArray(repos.flat());
+      } else if (config?.namespaces) {
+        logger.debug('Auto-discovering Gitea repositories by organization');
```
Comment (viceice) [inline lib/modules/platform/gitea/index.ts:327]:
```suggestion
        logger.debug({ namespaces: config.namespaces }, 'Auto-discovering by organization');
```

**Review by viceice — COMMENTED** [review]
Align logging to github

**Review by viceice — DISMISSED** [review]
we should refactor in a follow up PR to do the mirror, archive and deduplicate as a single last step, so we can deduplicate some more code.

**Review by viceice — DISMISSED** [review]

**Comment by rvanbutselaar** [comment]
I was wondering if we can merge this change. We would like to use this functionality to reduce the load on Gitea.

**Comment by rarkins** [comment]
Conflicts need resolving first

**Review by viceice — APPROVED** [review]

**Pull request merged**